sql/src/main/java/io/crate/planner/Merge.java

```diff
@@ -22,20 +22,144 @@
 
     private final Plan subPlan;
     private final MergePhase mergePhase;
 
-    public Merge(Plan subPlan, MergePhase mergePhase) {
+    private int limit;
+    private int offset;
+    private int numOutputs;
+
+    @Nullable
+    private PositionalOrderBy orderBy;
+
+    public static Merge create(Plan subPlan,
+                               Planner.Context context,
+                               List<Projection> projections,
+                               int limit,
+                               int offset,
+                               int numOutputs) {
+        MergePhase mergePhase = new MergePhase(
+            context.jobId(),
+            context.nextExecutionPhaseId(),
+            "mergeOnHandler",
+            subPlan.resultDescription().nodeIds().size(),
+            Collections.emptyList(),
+            subPlan.resultDescription().streamOutputs(),
+            projections,
+            DistributionInfo.DEFAULT_SAME_NODE,
+            null
+        );
+        return new Merge(subPlan, mergePhase, limit, offset, numOutputs, null);
+    }
+
+    /**
+     * Wrap the subPlan into a Merge plan if it isn't executed on the handler.
+     * @param projections projections to be applied on the subPlan or merge plan.
+     *                    These projections must not affect the limit, offset, order by or numOutputs
+     *
+     *                    If the subPlan contains a limit/offset or orderBy in its resultDescription this method will
+     *                    add a TopNProjection AFTER the projections.
+     */
+    public static Plan mergeToHandler(Plan subPlan, Planner.Context plannerContext, List<Projection> projections) {
+        ResultDescription resultDescription = subPlan.resultDescription();
+        assert resultDescription != null : "all plans must have a result description. Plan without: " + subPlan;
+
+        // If a sub-Plan applies a limit it is usually limit+offset
+        // So even if the execution is on the handler the limit may be too large and the final limit needs to be applied as well
+        TopNProjection topN = TopNProjection.createIfNeeded(
+            resultDescription.limit(),
+            resultDescription.offset(),
+            resultDescription.numOutputs(),
+            resultDescription.streamOutputs());
+        if (ExecutionPhases.executesOnHandler(plannerContext.handlerNode(), resultDescription.nodeIds())) {
+            for (Projection projection : projections) {
+                assert projection.outputs().size() == resultDescription.numOutputs()
+                    : "projection must not affect numOutputs";
+                subPlan.addProjection(projection, null, null, null, null);
+            }
+            if (topN != null) {
+                subPlan.addProjection(topN, TopN.NO_LIMIT, 0, resultDescription.numOutputs(), null);
+            }
+            // resultDescription.orderBy can be ignored here because it is only relevant to do a sorted merge
```

Comment:
the main factor for not looking at the order is that there is only one bucket ... pls mention